### DISCUSSION TITLE
Insertion on nested structures with a known base type and different fields

### DISCUSSION THREAD

**Discussion by isidentical** [geldata/gel#2445] [Q&A] [answered]
Since it is not possible to insert 2 same types within same query (due to this being explicitly forbidden), was wondering whether is there an efficient way of inserting algebraic datatypes (eg ASTs).

I currently insert a tree in multiple queries from bottom leafs up to the root node, where as an example for `2 + 2`, I do (in order) `obj1 = (INSERT Constant {value := 2})`, `obj2 = (INSERT Constant {value := 2})`, `(INSERT Add {left := (SELECT expr FILTER .id = obj1.id), right := (SELECT expr FILTER .id = obj2.id)})` (where `expr` is the base type of `Constant` and `Add`, and `Add.left/right` are `expr`s). 

However this becomes extremely slow, and might reach 5-6 seconds per tree (even more on highly complex ASTs). I know that I can also do `SELECT DETACHED`, but that still means I have to insert the children of the node first. Is there a way to do these operations in a rather fast way, or is this the optimal form?

**Comment by elprans** [comment, marked as answer]
You don't need to filter by id to refer to the inserted objects, just refer to them directly:

```
WITH
    obj1 := (INSERT Constant {value := 2}), 
    obj2 := (INSERT Constant {value := 2}), 
INSERT Add {left := obj1, right := obj2}
```

**Reply by isidentical** [reply]
I wasn't aware of this, thanks! 

Though it appears to be this form gets exponentially slower when executing bigger queries compared the other solution (inserting `obj1` and `obj2` separately and then selecting references when inserting the parent). I don't know whether this has something to do with the underlying edgeql compiler or maybe the input being giant though here is an [example query](https://gist.github.com/isidentical/a2bbd3ff53c5ffac4f32d5ade0620e17) that takes `120` seconds to execute (previously it was `4.7` seconds), and [here is a relatively small one](https://gist.github.com/isidentical/46124bd49680b992d78b56266c5038ab) that takes `4.8` seconds (used to take only `1.1` second to execute).  

It might be a bit silly question (considering these queries might not make sense) but is there any potential optimization ideas that you can give me? Or is there any way of inserting static data to the EdgeDB (like in a bulk insert manner, from json or some sort of structured dump).

**Reply by elprans** [reply]
Yes, such large amount of EdgeQL would make query compilation a bottleneck, but also I imagine Postgres isn't all too happy about several megabytes of SQL in a single statement.

> but is there any potential optimization ideas that you can give me

Since the issue here is the size of the query and its uncacheability, I'd suggest reworking the generator to do depth-first insertion in multiple statements.  Individual statements would cache well both on EdgeDB and Postgres side.  The downside is extra latency for server roundtrips.  To optimize further, you can batch insertions of elements of same node type that don't have cross-dependencies by packing them into JSON arrays:

```
WITH 
    constants := <array<tuple<
                     value: str,
                     lineno: int32,
                     col_offset: int32,
                     end_lineno: int32,
                     end_col_offset: int32,
                 >>> <json> $constants,

FOR const IN {array_unpack(constants}}
UNION (
    INSERT Constant {
        unique_id := const.unique_id,
        value := const.value,
        ...
    }
)
```

> Or is there any way of inserting static data to the EdgeDB (like in a bulk insert manner, from json or some sort of structured dump).

We have the semi-structured JSON import support on the roadmap, and your use case could be a good proving ground for it too!  Efficient insertion of arbitrary trees representing data for a cyclic type graph isn't a trivial problem, but we have some ideas no how to tackle this.

**Reply by isidentical** [reply]
Thanks so much, I'll see what I can do related to the second statement and will be looking forward to the third one :eyes:
